library/src/main/kotlin/com/livefront/gsonkotlinadapter/KoltinReflectiveTypeAdapterFactory.kt

```diff
@@ -40,61 +40,55 @@ class KotlinReflectiveTypeAdapterFactory private constructor(
-
-        /**
-         * Provides a mapping between parameter and automatic type-specific defaults (i.e. null for
-         * objects, 0 for numeric primitives, '\u0000' for chars, false for booleans). If the
-         * parameter [KParameter.isOptional] then it will not exist in this map and the
-         * manually-supplied default value will be used when the constructor is called. The
-         * contents of this map may vary based on the [enableDefaultPrimitiveValues] property.
-         */
-        private val constructorParameterDefaultsMap: Map<KParameter, Any?> = primaryConstructor
-            .parameters
-            .filterNot(KParameter::isOptional)
-            .associateWith { if (enableDefaultPrimitiveValues) it.defaultValue else null }
 
-        private val constructorParameterNameMap: Map<KParameter, List<String>> = primaryConstructor
-            .parameters
-            .map { it to it.getSerializedNames(declaringClass) }
-            .toMap()
+        val primaryConstructor: KFunction<T> = kotlinRawType
+            .primaryConstructor
+            ?.apply { isAccessible = true }
+            ?: return null
+        val declaringClass: Class<T> = primaryConstructor.javaConstructor!!.declaringClass!!
+        val innerAdapters: MutableMap<KParameter, TypeAdapter<*>> = mutableMapOf()
+        val constructorParameterDefaultsMap: MutableMap<KParameter, Any?> = mutableMapOf()
+        val invalidReadParameters: MutableSet<KParameter> = mutableSetOf()
+        val constructorMap: MutableMap<String, KParameter> = mutableMapOf()
 
-        private val invalidReadParameters: Set<KParameter> = constructorParameterNameMap
-            .filter { (parameter, names) ->
-                if (enableDefaultPrimitiveValues && parameter.isPrimitive) {
-                    false
+        primaryConstructor.parameters.forEach { parameter: KParameter ->
+            val names: List<String> = parameter.getSerializedNames(declaringClass)
+            if (names.isNotEmpty()) {
+                innerAdapters[parameter] = gson.getAdapter(type.resolveParameterType(parameter))
+            }
+            if (!parameter.isOptional) {
+                constructorParameterDefaultsMap[parameter] = if (enableDefaultPrimitiveValues) {
+                    parameter.defaultValue
                 } else {
-                    names.isEmpty() && !parameter.isOptional
+                    null
+                }
+                if (names.isEmpty() && (!parameter.isPrimitive || !enableDefaultPrimitiveValues)) {
+                    invalidReadParameters += parameter
                 }
             }
-            .keys
-
-        private val constructorMap: Map<String, KParameter> = constructorParameterNameMap
-            .entries
-            .flatMap { (parameter, names) -> names.map { it to parameter } }
-            .toMap()
+            names.forEach { name: String -> constructorMap[name] = parameter }
+        }
```

Comment:
The only comment I have is a general one, that I'd love to see more comments in this section due to it being such a condensed imperative block. That could mean some KDocs on the mutable maps declarations, as well as some explanations for each of the `if` statements.

Reply:
Yeah that could be nice too.